hw13

irbis45 opened the pull request:
Событие и обработчики на создание новости
Отправка на почту, в лог и телеграм

irbis45 committed:
hw13
# Событие и обработчики на создание новости
# Отправка на почту, в лог и телеграм

bravik commented on app/Events/NewsPublished.php:
```diff
@@ -0,0 +1,39 @@
+{
+    use Dispatchable, InteractsWithSockets, SerializesModels;
+
+    public NewsDTO $news;
```
Не рекоммендую в событиях использовать любые модели, DTO из прочих слоев. В событиях идеально будет если будут примитивные данные PHP. События  стоит воспринимать как DTO-шки. Они не должны содержать сложных структур. Смысл в том, что они должны сериализовываться в легковесные универсальные структуры вроде JSON, желательно без привязки к классам. Если вы даже не используете несколько сервисов, представьте, что вы выши класссы сериализуете как по умолчанию ларавель делает и у вас в проде возникает большое количество таких событий. И вы класс отрефакторили и нейминг или структура изменилась - все ваша сериализация сломана и се события в очереди после деплоя не смогут быть обработаны

bravik commented on app/Http/Controllers/Admin/News/CreateController.php:
```diff
@@ -38,15 +42,27 @@ public function create(AuthManager $authManager): View
 
-        $createNewsUseCase(new CommandDTO($request->get('title'), $request->get('content'), $userId, $request->get('category_id'), $request->get('published_at'), $request->get('is_draft', false)), $isAdmin);
+        try {
+            $newsId = $createNewsUseCase(new CommandDTO($request->get('title'), $request->get('content'), $userId, $request->get('category_id'), $request->get('published_at'), $request->get('is_draft', false)), $isAdmin);
```
переносы строк сделали бы код читабельнее

bravik commented on app/Services/User/Handlers/GetSubscribedNewsHandler.php:
```diff
@@ -0,0 +1,21 @@
+use App\Services\User\Repositories\UserRepositoryInterface;
+use App\Services\User\Results\Fetcher;
+
+class GetSubscribedNewsHandler
```
Handler-ом обычно называется команда, у вас тут запрос, больше подходит Fetcher. 
Кроме того у вас есть еще и вложенный фетчер.. 
Не рекомендую команды и запросы использовать внутри друг друга, обычно это атомарная единица действия. Если какую-то логику надо переисползовать можно в отдельный сервис/helper вынести её и использовать внутри комманд/запросов

bravik commented on app/Services/User/Repositories/UserRepositoryInterface.php:
```diff
@@ -54,4 +54,9 @@ public function delete(User $user): ?bool;
     public function findByIds(array $ids): array;
+
+    /**
+     * @return array
```
Всегда типизируем массивы в докблоках

bravik commented on app/Services/User/Results/Fetcher.php:
```diff
@@ -4,20 +4,22 @@
 use App\Models\User;
+use App\Services\User\Results\User as ResultsUser;
+
 class Fetcher
```
нейминг не дает никаких представлений о том, что именно должен запрос вытаскивать

bravik commented on app/Services/User/Results/Fetcher.php:
```diff
@@ -4,20 +4,22 @@
+     * @return UserItems|ResultsUser
      */
-    public function fetch(array|User $users): UsersDTO|UserDTO
+    public function fetch(array|User $users): UserItems|ResultsUser
```
Лучше разделить на два отдельный фетчера GetUser FetchUsers, а не мешать вот так логику

bravik replied:
А поччему вообще в фетчер уже модель User передается? фетчер же сам должен вытащить эти модели, а вы тут просто заворачиваете в DTO модель.. это что-то непонятное сосем. Это не фетчер а UserToDTOMapper какой-то

bravik commented on app/Services/User/Results/User.php:
```diff
@@ -5,12 +5,13 @@
 use Carbon\Carbon;
 
-final readonly class UserDTO
+final readonly class User
```
Тут я бы оставил постфикс DTO, чтобы не путать с моделью

irbis45 committed:
hw13
# Правки посли review

bravik commented on app/Events/NewsPublished.php:
```diff
@@ -15,14 +14,11 @@ class NewsPublished
      * Create a new event instance.
      */
-    public function __construct(NewsDTO $news)
+    public function __construct(public int $id, public string $title, public string $content)
```
если методы public - следует добавить `readonly` на класс или на поля

bravik commented on app/Services/News/Handlers/CreateHandler.php:
```diff
@@ -15,10 +15,11 @@ public function __construct(private NewsRepositoryInterface $newsRepository)
+     * @return int|false
      */
-    public function __invoke(CommandDTO $commandDTO, bool $isAdmin = false): bool {
+    public function __invoke(CommandDTO $commandDTO, bool $isAdmin = false): int|false {
```
isAdmin уместнее в команду перенести...

bravik commented on app/Services/News/Handlers/CreateHandler.php:
```diff
@@ -33,11 +34,13 @@ public function __invoke(CommandDTO $commandDTO, bool $isAdmin = false): bool {
 
-        $res = $this->newsRepository->save($news);
+        if ($this->newsRepository->save($news)) {
+            Cache::tags('news')->flush(); // Очистить все кэши с тегом 'news'
```
Команды и их handler  - часть слоя с бизнес логикой.. Они не должны иметь конкретных инфраструктурных зависимостей, вместо этого определять некий интерфейс `CacheInterface`. А у вас не просто инфраструктурная зависимость, так еще и через фасад прямо минуя DI... Я еще пойму  если вы используете фасады где-то в контроллерах, которые по определению фреймворко-зависимые, но в слоях с бизнес логикой такого быть не должно

bravik commented on app/Services/User/Fetchers/FetchUsers.php:
```diff
@@ -0,0 +1,35 @@
+    /**
+     * @return UsersDTO
+     */
+    public function fetch(): UsersDTO
```
Я бы рекомендовал группировать сервисы и DTO-шки к ним относящиеся рядом. Но дело за вами конечно.. То есть
Commands/CreateUser/
 - Command
 - Handler
 
 Queries/FetchUsers
 - Query
 - Fetcher
 - Result (UsersDTO)
 - UserDTO

irbis45 replied:
спасибо за пояснения. Постараюсь учесть

irbis45 committed:
hw13
# Переработал структуру
# Добавил функционал из других ДЗ, которые не вошли в эту ветку изначально

irbis45 committed:
hw13
# scope'ы
# активность категорий
